contributors/design-proposals/storage/container-storage-interface-inline-volumes.md

```diff
@@ -0,0 +1,148 @@
+# In-line CSI volumes in Pods
+
+Author: @jsafrane
+
+## Goal
+* Define API and high level design for in-line CSI volumes in Pod
+
+## Motivation
+Currently, CSI can be used only though PersistentVolume object. All other persistent volume sources support in-line volumes in Pods, CSI should be no exception. There are two main drivers:
+* We want to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. In-line volumes should use CSI too.
+* CSI drivers can be used to provide Secrets-like volumes to pods, e.g. providing secrets from a remote vault. We don't want to force users to create PVs for each secret, we should allow to use them in-line in pods as regular Secrets or Secrets-like Flex volumes.
+
+## API
+`VolumeSource` needs to be extended with CSI volume source:
+```go
+type VolumeSource struct {
+    // <snip>
+
+	// CSI (Container Storage Interface) represents storage that handled by an external CSI driver (Beta feature).
+	// +optional
+	CSI *CSIVolumeSource
+}
+
+
+// Represents storage that is managed by an external CSI volume driver (Beta feature)
+type CSIVolumeSource struct {
+	// Driver is the name of the driver to use for this volume.
+	// Required.
+	Driver string
+
+	// VolumeHandle is the unique volume name returned by the CSI volume
+	// plugin’s CreateVolume to refer to the volume on all subsequent calls.
+	// Required.
```

Comment:
There needs to be a mechanism to make this optional. CSI allows a driver to not implement CreateVolume, in which case VolumeHandle could be generated by kubelet itself.

Reply:
ah. I was confusing the two. even still, the use case I see for one of the drivers looks like the following. There isn't really any benefit in requiring VolumeHandle I think:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - name: nginx
    image: nginx:1.13-alpine
    ports:
    - containerPort: 80
    volumeMount:
    - name: data
      mountPath: /usr/share/nginx/html
  volumes:
  - name: data
    csi:
      driver: drivers.k8s.io/image-driver
      image: kfox1111/misc:test
```

Reply:
There is no `CSIVolumeSource.Image`. If `kfox1111/misc:test` is unique volume ID (and it looks like it uniquely identifies source of the data to mount), it should be used as `VolumeHandle`.

Reply:
Hmm... I haven't figured out the actual mapping to k8s. Reading through the pv->csi api spec, I think it actually should be:
volumes:
  - name: data
    csi:
      driver: drivers.k8s.io/image-driver
      volumeAttributes:
        image: kfox1111/misc:test

I have a prototype csi driver working here: https://github.com/kfox1111/drivers/blob/volume-driver/pkg/image/nodeserver.go#L143

It uniquely identifies the container image to use for the volume. the volume identifier though should match the pod's lifecycle. so the easiest thing I can think of is to use the podid as the volumehandle. or maybe just a random id (but consistent for the pods lifecycle).

For things that follow pod lifecycle, VolumeHandle doesn't really make sense for the user to specify. It is supposed to be unique, but you could have multiple instances of the same pod on the same host. If the user specified something, then it could conflict between the pods on the same host.

Reply:
actually, thinking about it a little more, lets say you make a volume driver that was as trivial as possible. say, emptyDir for sake of discussion. Now say you had two instances of it within a pod:

```yaml
volumes:
  - name: foo
    csi:
      driver: drivers.k8s.io/emptyDir-driver
  - name: bar
    csi:
      driver: drivers.k8s.io/emptyDir-driver
 ```

Each instance would need to be unique per pod. so using pod_id as volume_id wouldn't be the right thing to do. Each should get its own volume_id. so maybe just randomly created ones might be the best option.

Drivers still might want to have access to the pod_id though. So maybe that could be passed through via an automatic volumeAttribute?

Reply:
Where does the VolumeHandlePrefix come from? is it specified in the pod spec? Should it come from the driver registrar (https://github.com/kubernetes-csi/driver-registrar) sidecar instead? Then the packager of the csi driver could tell k8s which type it expects so the user won't have to specify it most of the time.

Reply:
VolumeHandlePrefix comes in pod spec, `pod.spec.volumes[x].csi.volumeHandlePrefix`.

Reply:
So under my current understanding of the spec, the theoretical csi emtpyDir example would need to be:
```yaml
volumes:
  - name: foo
    csi:
      driver: drivers.k8s.io/emptyDir-driver
      volumeHandle: foo
      volumeHandlePrefix: pod
  - name: bar
    csi:
      driver: drivers.k8s.io/emptyDir-driver
      volumeHandle: bar
      volumeHandlePrefix: pod
```
Thats quite verbose. And a miss copied entry could have bad results. It really feels like volumeHandlePrefix or at least its default should come from the driver some how.

Alternately, could the VolumeHandlePrefix be made optional and default to Pod (should be a safe default) and VolumeHandle default to the volume's name and be optional (guaranteed to be unique within a pod). That would restore the ease of use for using any csi driver where you want to map volume lifecycle to the pod. and still allow for specifying the exact volumes you want to mount when you need to.